## “Does the identification of panic/recovery episodes change dramatically when reasonable nearby thresholds are used?”

The sensitivity test concerns the panic/recovery classification, not structural-break detection. Use the already-finalized regime data and rerun identify_panic_episodes() with different thresholds.

Panic threshold:          40%, 50%, 60%

Stabilization threshold:  20%, 25%, 30%


→ 3 × 3 = 9 specifications

Baseline: 50% panic / 25% stabilization

For each of the 9 combinations, calculate at least:
Number of completed episodes
Mean RecoveryDays
Median RecoveryDays
Number of assets with ≥1 episode

1. S&P 500 threshold sensitivity
   - 9 threshold combinations
   - useful for comparison with your original 03c development

2. Multi-asset threshold sensitivity
   - same 9 combinations
   - MAIN robustness analysis for the final methodology

# 03h — Panic and Stabilization Threshold Sensitivity Analysis

This notebook evaluates the sensitivity of the panic/recovery episode definition to alternative threshold choices.

The structural-break and EGARCH results are kept fixed. The analysis therefore uses the finalized volatility regimes obtained previously and varies only:

- the panic threshold;
- the stabilization threshold.

Baseline specification:

- Panic threshold = 50%
- Stabilization threshold = 25%

Sensitivity grid:

- Panic threshold: 40%, 50%, 60%
- Stabilization threshold: 20%, 25%, 30%

The analysis is performed first for the single-index S&P 500 specification and then for the final 22-asset dataset.

In [ ]:
from getpass import getpass

github_token = getpass("Enter your GitHub Token: ")
github_username = input("Enter your GitHub username: ")

!git clone https://{github_username}:{github_token}@github.com/{github_username}/Scientific-Project-SS26.git

%cd /content/Scientific-Project-SS26

Enter your GitHub Token: ··········
Enter your GitHub username: fleurclochette
Cloning into 'Scientific-Project-SS26'...
remote: Enumerating objects: 516, done.
remote: Counting objects: 100% (210/210), done.
remote: Compressing objects: 100% (150/150), done.
remote: Total 516 (delta 148), reused 85 (delta 60), pack-reused 306 (from 2)
Receiving objects: 100% (516/516), 13.13 MiB | 16.28 MiB/s, done.
Resolving deltas: 100% (261/261), done.
/content/Scientific-Project-SS26


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from itertools import product

pd.set_option("display.max_columns", None)
pd.set_option("display.width", 200)

print("Imports complete.")

Imports complete.


## Part A — Multi-asset sensitivity analysis

In [ ]:
multi_regimes = pd.read_csv(
    "data/processed/final_regimes_multi_asset_2000_2026.csv",
    parse_dates=["StartDate", "EndDate"]
)

multi_volatility = pd.read_csv(
    "data/processed/conditional_volatility_multi_asset_2000_2026.csv",
    parse_dates=["Date"]
)

baseline_episodes = pd.read_csv(
    "data/processed/final_panic_episodes_multi_asset_2000_2026.csv",
    parse_dates=["PanicStart", "StableDate"]
)

print("Regime rows:", len(multi_regimes))
print("Volatility rows:", len(multi_volatility))
print("Baseline episodes:", len(baseline_episodes))
print("Assets:", multi_regimes["Asset"].nunique())

Regime rows: 562
Volatility rows: 141615
Baseline episodes: 112
Assets: 22


In [ ]:
print("Regime columns:")
print(multi_regimes.columns.tolist())

print("\nVolatility columns:")
print(multi_volatility.columns.tolist())

print("\nEpisode columns:")
print(baseline_episodes.columns.tolist())

Regime columns:
['Asset', 'Regime', 'StartIndex', 'EndIndex', 'StartDate', 'EndDate', 'TradingDays', 'MeanVolatility', 'MedianVolatility', 'StdVolatility', 'MeanVolatilityPctChange']

Volatility columns:
['Date', 'Asset', 'LogReturn', 'ConditionalVolatility']

Episode columns:
['Asset', 'PanicRegime', 'PanicStart', 'PanicStartMeanVolatility', 'PeakRegime', 'PeakVolatility', 'StableRegime', 'StableDate', 'RecoveryDays']


In [ ]:
# Use the same episode identification function as 03g
def identify_panic_episodes(
    regime_df,
    volatility_dates,
    panic_threshold=0.50,
    stabilization_threshold=0.25
):
    episodes = []

    in_panic = False
    panic_regime = None
    panic_start = None
    panic_start_mean = None
    peak_regime = None
    peak_volatility = None

    regime_df = (
        regime_df
        .sort_values("Regime")
        .reset_index(drop=True)
    )

    volatility_dates = pd.DatetimeIndex(
        pd.to_datetime(volatility_dates)
    ).sort_values()

    for i in range(1, len(regime_df)):

        previous_mean = regime_df.loc[i - 1, "MeanVolatility"]
        current_mean = regime_df.loc[i, "MeanVolatility"]

        increase = (
            current_mean - previous_mean
        ) / previous_mean

        if not in_panic:

            if increase >= panic_threshold:

                in_panic = True

                panic_regime = int(
                    regime_df.loc[i, "Regime"]
                )

                panic_start = pd.to_datetime(
                    regime_df.loc[i, "StartDate"]
                )

                panic_start_mean = current_mean

                peak_regime = panic_regime
                peak_volatility = current_mean

        else:

            if current_mean > peak_volatility:

                peak_volatility = current_mean

                peak_regime = int(
                    regime_df.loc[i, "Regime"]
                )

            decline_from_peak = (
                peak_volatility - current_mean
            ) / peak_volatility

            if decline_from_peak >= stabilization_threshold:

                stable_regime = int(
                    regime_df.loc[i, "Regime"]
                )

                stable_date = pd.to_datetime(
                    regime_df.loc[i, "StartDate"]
                )

                panic_start_idx = volatility_dates.searchsorted(
                    panic_start
                )

                stable_idx = volatility_dates.searchsorted(
                    stable_date
                )

                recovery_days = (
                    stable_idx - panic_start_idx
                )

                episodes.append({
                    "PanicRegime": panic_regime,
                    "PanicStart": panic_start,
                    "PanicStartMeanVolatility": panic_start_mean,
                    "PeakRegime": peak_regime,
                    "PeakVolatility": peak_volatility,
                    "StableRegime": stable_regime,
                    "StableDate": stable_date,
                    "RecoveryDays": recovery_days
                })

                in_panic = False
                panic_regime = None
                panic_start = None
                panic_start_mean = None
                peak_regime = None
                peak_volatility = None

    return pd.DataFrame(episodes)

In [ ]:
# define the sensitivity grid
panic_thresholds = [0.40, 0.50, 0.60]
stabilization_thresholds = [0.20, 0.25, 0.30]

threshold_combinations = list(
    product(
        panic_thresholds,
        stabilization_thresholds
    )
)

threshold_combinations

[(0.4, 0.2),
 (0.4, 0.25),
 (0.4, 0.3),
 (0.5, 0.2),
 (0.5, 0.25),
 (0.5, 0.3),
 (0.6, 0.2),
 (0.6, 0.25),
 (0.6, 0.3)]

In [ ]:
# Run all 9 across all 22 assets
sensitivity_summaries = []
sensitivity_episodes = []

for panic_threshold, stabilization_threshold in threshold_combinations:

    specification_episodes = []

    for asset in sorted(multi_regimes["Asset"].unique()):

        asset_regimes = (
            multi_regimes[
                multi_regimes["Asset"] == asset
            ]
            .copy()
            .sort_values("Regime")
            .reset_index(drop=True)
        )

        asset_volatility_dates = (
            multi_volatility.loc[
                multi_volatility["Asset"] == asset,
                "Date"
            ]
            .sort_values()
        )

        episodes = identify_panic_episodes(
            regime_df=asset_regimes,
            volatility_dates=asset_volatility_dates,
            panic_threshold=panic_threshold,
            stabilization_threshold=stabilization_threshold
        )

        if not episodes.empty:

            episodes.insert(0, "Asset", asset)

            episodes["PanicThreshold"] = panic_threshold
            episodes["StabilizationThreshold"] = stabilization_threshold

            specification_episodes.append(episodes)

    if specification_episodes:

        specification_df = pd.concat(
            specification_episodes,
            ignore_index=True
        )

        sensitivity_episodes.append(specification_df)

        sensitivity_summaries.append({
            "PanicThreshold": panic_threshold,
            "StabilizationThreshold": stabilization_threshold,
            "NumberOfEpisodes": len(specification_df),
            "AssetsWithEpisodes": specification_df["Asset"].nunique(),
            "MeanRecoveryDays": specification_df["RecoveryDays"].mean(),
            "MedianRecoveryDays": specification_df["RecoveryDays"].median(),
            "MinRecoveryDays": specification_df["RecoveryDays"].min(),
            "MaxRecoveryDays": specification_df["RecoveryDays"].max()
        })

    else:

        sensitivity_summaries.append({
            "PanicThreshold": panic_threshold,
            "StabilizationThreshold": stabilization_threshold,
            "NumberOfEpisodes": 0,
            "AssetsWithEpisodes": 0,
            "MeanRecoveryDays": np.nan,
            "MedianRecoveryDays": np.nan,
            "MinRecoveryDays": np.nan,
            "MaxRecoveryDays": np.nan
        })

multi_sensitivity_summary = pd.DataFrame(
    sensitivity_summaries
)

multi_sensitivity_episodes = pd.concat(
    sensitivity_episodes,
    ignore_index=True
)

display(multi_sensitivity_summary)

,PanicThreshold,StabilizationThreshold,NumberOfEpisodes,AssetsWithEpisodes,MeanRecoveryDays,MedianRecoveryDays,MinRecoveryDays,MaxRecoveryDays
0,0.4,0.20,128,22,162.984375,102.5,60,1306
1,0.4,0.25,124,22,174.153226,107.0,60,803
2,0.4,0.30,117,22,206.683761,115.0,60,1314
3,0.5,0.20,114,22,150.640351,93.5,60,1306
4,0.5,0.25,112,22,148.723214,95.5,60,614
5,0.5,0.30,106,22,188.188679,108.0,60,1314
6,0.6,0.20,95,22,144.936842,88.0,60,1306
7,0.6,0.25,93,22,138.838710,88.0,60,614
8,0.6,0.30,89,22,161.842697,102.0,60,744


In [ ]:
# make percentages easier to read
multi_sensitivity_display = (
    multi_sensitivity_summary.copy()
)

multi_sensitivity_display["PanicThreshold"] = (
    multi_sensitivity_display["PanicThreshold"] * 100
)

multi_sensitivity_display["StabilizationThreshold"] = (
    multi_sensitivity_display["StabilizationThreshold"] * 100
)

multi_sensitivity_display = (
    multi_sensitivity_display
    .sort_values(
        ["PanicThreshold", "StabilizationThreshold"]
    )
    .reset_index(drop=True)
)

display(multi_sensitivity_display.round(2))

,PanicThreshold,StabilizationThreshold,NumberOfEpisodes,AssetsWithEpisodes,MeanRecoveryDays,MedianRecoveryDays,MinRecoveryDays,MaxRecoveryDays
0,40.0,20.0,128,22,162.98,102.5,60,1306
1,40.0,25.0,124,22,174.15,107.0,60,803
2,40.0,30.0,117,22,206.68,115.0,60,1314
3,50.0,20.0,114,22,150.64,93.5,60,1306
4,50.0,25.0,112,22,148.72,95.5,60,614
5,50.0,30.0,106,22,188.19,108.0,60,1314
6,60.0,20.0,95,22,144.94,88.0,60,1306
7,60.0,25.0,93,22,138.84,88.0,60,614
8,60.0,30.0,89,22,161.84,102.0,60,744


In [ ]:
# Verify that 50% / 25% reproduces the final 112 episodes
baseline_check = multi_sensitivity_summary[
    (multi_sensitivity_summary["PanicThreshold"] == 0.50) &
    (multi_sensitivity_summary["StabilizationThreshold"] == 0.25)
]

display(baseline_check)

baseline_count = int(
    baseline_check["NumberOfEpisodes"].iloc[0]
)

print("Sensitivity baseline episode count:", baseline_count)
print("Saved 03g baseline episode count:", len(baseline_episodes))
print("Counts match:", baseline_count == len(baseline_episodes))

,PanicThreshold,StabilizationThreshold,NumberOfEpisodes,AssetsWithEpisodes,MeanRecoveryDays,MedianRecoveryDays,MinRecoveryDays,MaxRecoveryDays
4,0.5,0.25,112,22,148.723214,95.5,60,614


Sensitivity baseline episode count: 112
Saved 03g baseline episode count: 112
Counts match: True


In [ ]:
# Now check stability of the episode identification
# We also want to know whether the same panic episodes survive nearby threshold choices
# Compare panic starts against the 50/25 baseline
baseline_keys = set(
    zip(
        baseline_episodes["Asset"],
        baseline_episodes["PanicStart"]
    )
)

overlap_results = []

for panic_threshold, stabilization_threshold in threshold_combinations:

    current = multi_sensitivity_episodes[
        (multi_sensitivity_episodes["PanicThreshold"] == panic_threshold) &
        (
            multi_sensitivity_episodes["StabilizationThreshold"]
            == stabilization_threshold
        )
    ]

    current_keys = set(
        zip(
            current["Asset"],
            current["PanicStart"]
        )
    )

    intersection = len(
        baseline_keys.intersection(current_keys)
    )

    union = len(
        baseline_keys.union(current_keys)
    )

    jaccard = (
        intersection / union
        if union > 0
        else np.nan
    )

    baseline_retained = (
        intersection / len(baseline_keys)
        if len(baseline_keys) > 0
        else np.nan
    )

    overlap_results.append({
        "PanicThreshold": panic_threshold,
        "StabilizationThreshold": stabilization_threshold,
        "Episodes": len(current),
        "BaselineEpisodesAlsoIdentified": intersection,
        "BaselineEpisodeRetention": baseline_retained,
        "JaccardSimilarity": jaccard
    })

overlap_df = pd.DataFrame(overlap_results)

overlap_df["PanicThresholdPct"] = (
    overlap_df["PanicThreshold"] * 100
)

overlap_df["StabilizationThresholdPct"] = (
    overlap_df["StabilizationThreshold"] * 100
)

display(
    overlap_df[
        [
            "PanicThresholdPct",
            "StabilizationThresholdPct",
            "Episodes",
            "BaselineEpisodesAlsoIdentified",
            "BaselineEpisodeRetention",
            "JaccardSimilarity"
        ]
    ].round(3)
)

,PanicThresholdPct,StabilizationThresholdPct,Episodes,BaselineEpisodesAlsoIdentified,BaselineEpisodeRetention,JaccardSimilarity
0,40.0,20.0,128,104,0.929,0.765
1,40.0,25.0,124,102,0.911,0.761
2,40.0,30.0,117,96,0.857,0.722
3,50.0,20.0,114,112,1.000,0.982
4,50.0,25.0,112,112,1.000,1.000
5,50.0,30.0,106,106,0.946,0.946
6,60.0,20.0,95,90,0.804,0.769
7,60.0,25.0,93,90,0.804,0.783
8,60.0,30.0,89,87,0.777,0.763


## Part B — Single S&P robustness

In [ ]:
sp500_regimes = pd.read_csv(
    "data/processed/final_regimes_egarch_2000_2026.csv",
    parse_dates=["StartDate", "EndDate"]
)

sp500_volatility = pd.read_csv(
    "data/processed/conditional_volatility_egarch_2000_2026.csv",
    parse_dates=["Date"]
)

sp500_baseline_episodes = pd.read_csv(
    "data/processed/final_panic_episodes_egarch_2000_2026.csv",
    parse_dates=["PanicStart", "StableDate"]
)

print("S&P regimes:", len(sp500_regimes))
print("S&P volatility observations:", len(sp500_volatility))
print("S&P baseline episodes:", len(sp500_baseline_episodes))

S&P regimes: 21
S&P volatility observations: 6683
S&P baseline episodes: 5


In [ ]:
# Run the same 3×3 sensitivity analysis for S&P
sp500_sensitivity_results = []
sp500_sensitivity_episodes = []

sp500_volatility_dates = (
    sp500_volatility["Date"]
    .sort_values()
)

for panic_threshold, stabilization_threshold in threshold_combinations:

    episodes = identify_panic_episodes(
        regime_df=sp500_regimes.copy(),
        volatility_dates=sp500_volatility_dates,
        panic_threshold=panic_threshold,
        stabilization_threshold=stabilization_threshold
    )

    if not episodes.empty:

        episodes["PanicThreshold"] = panic_threshold
        episodes["StabilizationThreshold"] = stabilization_threshold

        sp500_sensitivity_episodes.append(episodes)

        sp500_sensitivity_results.append({
            "PanicThreshold": panic_threshold,
            "StabilizationThreshold": stabilization_threshold,
            "NumberOfEpisodes": len(episodes),
            "MeanRecoveryDays": episodes["RecoveryDays"].mean(),
            "MedianRecoveryDays": episodes["RecoveryDays"].median()
        })

    else:

        sp500_sensitivity_results.append({
            "PanicThreshold": panic_threshold,
            "StabilizationThreshold": stabilization_threshold,
            "NumberOfEpisodes": 0,
            "MeanRecoveryDays": np.nan,
            "MedianRecoveryDays": np.nan
        })

sp500_sensitivity_summary = pd.DataFrame(
    sp500_sensitivity_results
)

sp500_sensitivity_display = (
    sp500_sensitivity_summary.copy()
)

sp500_sensitivity_display["PanicThreshold"] *= 100
sp500_sensitivity_display["StabilizationThreshold"] *= 100

display(sp500_sensitivity_display.round(2))

,PanicThreshold,StabilizationThreshold,NumberOfEpisodes,MeanRecoveryDays,MedianRecoveryDays
0,40.0,20.0,6,183.33,99.5
1,40.0,25.0,6,183.33,99.5
2,40.0,30.0,6,203.33,156.5
3,50.0,20.0,5,181.40,108.0
4,50.0,25.0,5,181.40,108.0
5,50.0,30.0,5,205.40,205.0
6,60.0,20.0,4,205.50,204.5
7,60.0,25.0,4,205.50,204.5
8,60.0,30.0,4,205.50,204.5


In [ ]:
# check the S&P baseline
sp500_baseline_check = sp500_sensitivity_summary[
    (sp500_sensitivity_summary["PanicThreshold"] == 0.50) &
    (
        sp500_sensitivity_summary["StabilizationThreshold"]
        == 0.25
    )
]

display(sp500_baseline_check)

print(
    "Recalculated baseline:",
    int(sp500_baseline_check["NumberOfEpisodes"].iloc[0])
)

print(
    "Saved baseline:",
    len(sp500_baseline_episodes)
)

,PanicThreshold,StabilizationThreshold,NumberOfEpisodes,MeanRecoveryDays,MedianRecoveryDays
4,0.5,0.25,5,181.4,108.0


Recalculated baseline: 5
Saved baseline: 5


The baseline thresholds were selected to distinguish economically substantial volatility regime changes from smaller fluctuations while avoiding an excessively restrictive definition of market stress. A panic threshold of 50% represents an intermediate specification between the more permissive 40% and more restrictive 60% alternatives. Similarly, stabilization was defined as a 25% decline from peak regime volatility, lying between the tested 20% and 30% alternatives. A sensitivity analysis confirms that the baseline does not represent a fragile threshold choice. Holding the panic threshold at 50%, changing the stabilization threshold to 20% or 30% resulted in 114 and 106 episodes, respectively, compared with 112 under the baseline. Moreover, 100% and 94.6% of baseline panic starts were retained. Variation in the panic threshold had a larger effect, as expected, but substantial overlap with the baseline remained: 91.1% of baseline panic starts were retained under the 40%/25% specification and 80.4% under 60%/25%. Thus, the 50%/25% specification is used as an intermediate operational definition rather than because it maximizes the number of identified episodes.

A limitation of the episode-identification procedure is that the 50% panic and 25% stabilization thresholds are operational choices rather than statistically estimated or theoretically unique cutoffs. The sensitivity analysis shows that the identified episode set remains relatively stable under nearby threshold specifications, but this does not establish 50%/25% as an optimal definition of market panic and recovery. In particular, the panic threshold has a noticeable effect on the number of identified episodes. Consequently, the resulting recovery sample remains conditional on the chosen definition of market stress.

This limitation is mitigated by applying the same thresholds consistently across all assets and by explicitly evaluating alternative threshold specifications.